modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2432,56 +2433,58 @@ def _apply_func_to_range_partitioning(
+                        range(step, len(self._partitions), step),
+                    )
+
+                new_partitions = np.array(
+                    [
+                        self._partition_mgr_cls.column_partitions(
+                            ptn_grp, full_axis=False
+                        )
+                        for ptn_grp in joining_partitions
+                    ]
+                )
+            else:
+                new_partitions = self._partitions
         else:
-            # This means we are not sorting numbers, so we need our quantiles to not try
-            # arithmetic on the values.
-            method = "inverted_cdf"
+            new_partitions = self._partitions
 
-        shuffling_functions = build_sort_functions(
+        shuffling_functions = ShuffleSortFunctions(
             self,
-            key_column,
-            method,
+            key_columns,
             ascending[0] if is_list_like(ascending) else ascending,
             ideal_num_new_partitions,
             **kwargs,
         )
-        if ideal_num_new_partitions < len(self._partitions):
-            if len(self._partitions) % ideal_num_new_partitions == 0:
-                joining_partitions = np.split(
-                    self._partitions, ideal_num_new_partitions
-                )
-            else:
-                joining_partitions = np.split(
-                    self._partitions,
-                    range(
-                        0,
-                        len(self._partitions),
-                        round(len(self._partitions) / ideal_num_new_partitions),
-                    )[1:],
-                )
 
-            new_partitions = np.array(
-                [
-                    self._partition_mgr_cls.column_partitions(ptn_grp, full_axis=False)
-                    for ptn_grp in joining_partitions
-                ]
-            )
-        else:
-            new_partitions = self._partitions
+        # here we want to get indices of those partitions that hold the key columns;
+        # first we translate column labels into their numeric indices
+        key_indices = self.columns.get_indexer_for(key_columns)
+        # 'indices' will show us partition boundaries, helping to understand which
+        # column belongs to which partition. For example if 'indices = [0, 5, 10, 15]'
+        # then we know that columns with indices (0-4) are located in part#0;
+        # columns with indices (5-9) are located in part#1 and so on...
+        indices = np.cumsum([0] + self.column_widths)
+        # 'partition_indices' will store partition ids that hold the key columns
+        partition_indices = set()
+        for i in range(len(indices) - 1):
+            # going through the key columns and check whether they belong to the part#i
+            for key_idx in key_indices:
+                if key_idx >= indices[i] and key_idx < indices[i + 1]:
+                    partition_indices.add(i)
```

Comment:
Why not use `break` here?

Reply:
[solved](https://github.com/modin-project/modin/pull/6533#discussion_r1319906745)